### PR TITLE
ARROW-1750: [C++] Remove the need for arrow/util/random.h

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

**Comment by cpcloud** [comment]
@wesm @xhochy Can one of you review when you get a chance? Thanks!

File: cpp/src/arrow/array-test.cc [hunk 1]
```diff
@@ -272,6 +274,24 @@ class TestPrimitiveBuilder : public TestBuilder {
   vector<uint8_t> valid_bytes_;
 };
 
+/// \brief uint8_t isn't a valid template parameter to uniform_int_distribution, so
+/// we use SampleType to determine which kind of integer to use to sample.
+template <typename T,
+          typename = typename std::enable_if<std::is_integral<T>::value, T>::type>
+struct SampleType {
```
Comment (cpcloud) [inline cpp/src/arrow/array-test.cc]: I'm going to rename this to `UniformIntSampleType`.

File: cpp/src/arrow/test-util.h [hunk 2]
```diff
@@ -143,56 +140,104 @@ static inline Status GetBitmapFromVector(const std::vector<T>& is_valid,
...
+  DCHECK_GE(precision, 1) << "decimal precision must be greater than or equal to 1, got "
+                          << precision;
+  DCHECK_LE(precision, 38) << "decimal precision must be less than or equal to 38, got "
+                           << precision;
+  DCHECK_NE(min_decimal, NULLPTR);
+  DCHECK_NE(max_decimal, NULLPTR);
+
+  *max_decimal = 1;
+  for (int32_t i = 0; i < precision; ++i) {
+    *max_decimal *= 10;
```
Comment (cpcloud) [inline cpp/src/arrow/test-util.h]: After ARROW-1749 is merged, this will need to be rewritten.

**Review by wesm — COMMENTED** [review]
+1, pending rebase / green build

**Comment by cpcloud** [comment]
Not sure why, but the OSX build timed out on Travis CI.

**Comment by cpcloud** [comment]
@wesm This is passing here: https://travis-ci.org/cpcloud/arrow/builds/298006973

**Pull request closed without merging**